src/bp-members/classes/class-bp-members-notice.php

```diff
@@ -0,0 +1,758 @@
+		}
+
+		// The meta query.
+		if ( ! empty( $meta_query_sql['join'] ) ) {
+			$join_sql = $meta_query_sql['join'];
+		}
+
+		// 127 is the value used to deactivate a notice.
+		if ( 'active' === $r['type'] ) {
+			$where_conditions['type'] = 'n.priority != 127';
+		} elseif ( 'inactive' === $r['type'] ) {
+			$where_conditions['type'] = 'n.priority = 127';
+		}
+
+		if ( $r['exclude'] ) {
+			$where_conditions['exclude'] = 'n.id NOT IN( ' . implode( ', ', wp_parse_id_list( $r['exclude'] ) ) . ' )';
+		}
+
+		if ( $r['target__in'] ) {
+			$where_conditions['target__in'] = 'n.target IN( \'' . implode( '\', \'', wp_parse_slug_list( $r['target__in'] ) ) . '\' )';
+		}
+
+		if ( 'inactive' !== $r['type'] && ! is_null( $r['priority'] ) && is_numeric( $r['priority'] ) ) {
+			$where_conditions['priority'] = $wpdb->prepare( 'priority = %d', $r['priority'] );
+		}
+
+		if ( $r['user_id'] ) {
+			if ( ! bp_user_can( $r['user_id'], 'edit_posts' ) ) {
+				$where_conditions['user_cap'] = 'n.target = \'community\'';
+			} elseif ( ! bp_user_can( $r['user_id'], 'manage_options' ) ) {
+				$where_conditions['user_cap'] = 'n.target != \'admins\'';
+			}
+		}
+
+		$limit_sql = '';
+		if ( (int) $r['pag_num'] >= 0 ) {
+			$limit_sql = $wpdb->prepare( "LIMIT %d, %d", (int) ( ( $r['pag_page'] - 1 ) * $r['pag_num'] ), (int) $r['pag_num'] );
+		}
+
+		// Custom WHERE.
+		if ( ! empty( $where_conditions ) ) {
+			$where_sql = 'WHERE ' . implode( ' AND ', $where_conditions );
+		}
+
+		$bp = buddypress();
+
+		// Get notice IDs.
+		if ( 'ids' === $r['fields'] ) {
+			$result = $wpdb->get_col(
+				"SELECT n.id FROM {$bp->members->table_name_notices} n
+				{$join_sql}
+				{$where_sql}"
+			);
+			$result = wp_parse_id_list( $result );
+
+			// Get the first active notice.
+		} elseif ( $is_first_active ) {
+			$result = $wpdb->get_row(
+				"SELECT n.* FROM {$bp->members->table_name_notices} n
+				{$join_sql}
+				{$where_sql}
+				ORDER BY priority ASC, date_sent DESC
+				{$limit_sql}"
+			);
+
+			// Get the notices count.
+		} elseif ( true === $r['count_total_only'] ) {
+			$result = $wpdb->get_var(
+				"SELECT COUNT(*) FROM {$bp->members->table_name_notices} n
+				{$join_sql}
+				{$where_sql}"
+			);
+
+			// Get all matching notices.
+		} else {
+			$user_result     = false;
+			$is_user_notices = isset( $args['user_id'], $args['exclude'] ) && $args['user_id'] && 1 === $r['pag_page'] && 0 !== $r['priority'];
+
+			// Are we getting the user's top priority notices?
+			if ( $is_user_notices ) {
+				$user_result = wp_cache_get( $args['user_id'], 'bp_member_top_priority_notices' );
+			}
+
+			if ( false === $user_result ) {
+				$result = $wpdb->get_results(
+					"SELECT n.* FROM {$bp->members->table_name_notices} n
+					{$join_sql}
+					{$where_sql}
+					ORDER BY priority ASC, date_sent DESC
+					{$limit_sql}"
+				);
+
+				if ( $is_user_notices ) {
+					wp_cache_set( $args['user_id'], $result, 'bp_member_top_priority_notices' );
```

Comment:
As part of https://buddypress.trac.wordpress.org/ticket/8552, we should add support for `cache_results`. I can do it after this is merged too.

Reply:
I agree, let's do it once it's merged 👍